File: drive/src/main/java/de/mein/drive/sql/dao/TransferDao.java
package de.mein.drive.sql.dao;

import de.mein.drive.sql.*;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.sqlite.SQLiteErrorCode;
import org.sqlite.SQLiteException;

import de.mein.auth.data.db.MissingHash;
import de.mein.auth.tools.N;
import de.mein.drive.tasks.AvailHashEntry;
import de.mein.sql.*;

import org.sqlite.core.DB;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;

/**
 * Created by xor on 12/16/16.
 */
public class TransferDao extends Dao {
    private DbTransferDetails dummy = new DbTransferDetails();

    public TransferDao(ISQLQueries ISQLQueries) {
        super(ISQLQueries);
    }

    public TransferDao(ISQLQueries ISQLQueries, boolean lock) {
        super(ISQLQueries, lock);
    }

    public TransferLeftovers getLeftoversByService(Long partnerCertId, String partnerServiceUuid) throws SqlQueriesException {
        TransferLeftovers lo = new TransferLeftovers();
        String query = "select count(1) as " + lo.getFilesTotal().k()
                + ", (select count(1) from " + dummy.getTableName() + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and " + dummy.getState().k() + "=? ) as " + lo.getFilesTransferred().k()
                + ", sum(" + dummy.getSize().k() + ") as " + lo.getBytesTotal().k()
                + ", (select sum(" + dummy.getTransferred().k() + ") from " + dummy.getTableName() + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and " + dummy.getState().k() + " != ?) as " + lo.getBytesTransferred().k()
                + " from " + dummy.getTableName() + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and " + dummy.getState().k() + " !=?";
        List<TransferLeftovers> list = sqlQueries.loadString(lo.getAllAttributes(), lo, query, ISQLQueries.whereArgs(partnerCertId, partnerServiceUuid, TransferState.DONE,
                partnerCertId, partnerServiceUuid, TransferState.SUSPENDED,
                partnerCertId, partnerServiceUuid, TransferState.SUSPENDED));
        if (!list.isEmpty()) {
            // summed bytes are null if no transfer exists
            TransferLeftovers result = list.get(0);
            if (result.getBytesTransferred().equalsValue(null))
                result.getBytesTransferred().v(0L);
            if (result.getBytesTotal().equalsValue(null))
                result.getBytesTotal().v(0L);
            return result;
        }
        return null;
    }

    public void insert(DbTransferDetails dbTransferDetails) throws SqlQueriesException {
        try {
            Long id = sqlQueries.insert(dbTransferDetails);
            dbTransferDetails.getId().v(id);
        } catch (SqlQueriesException sqlQueriesException) {
            // non unique transfers happen quite often and thats normal. don't fill the logs with that.
            if (sqlQueriesException.getException() instanceof SQLiteException) {
                SQLiteException sqLiteException = (SQLiteException) sqlQueriesException.getException();
                if (sqLiteException.getResultCode() != SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE)
                    throw sqlQueriesException;

            }
        }

    }

    public DbTransferDetails getOneTransfer() throws SqlQueriesException {

        List<DbTransferDetails> res = sqlQueries.load(dummy.getAllAttributes(), dummy, null, null, " order by " + dummy.getId().k() + " limit 1");
        if (res.size() == 1)
            return res.get(0);
        return null;
    }

    public void delete(Long id) throws SqlQueriesException {

        sqlQueries.delete(dummy, dummy.getId().k() + "=?", ISQLQueries.whereArgs(id));
    }

    public List<DbTransferDetails> getTwoTransferSets() throws SqlQueriesException {
        return getTransferSets(2);
    }

    /**
     * @param limit
     * @return all TransferSets if limit is null
     */
    public List<DbTransferDetails> getTransferSets(Integer limit) throws SqlQueriesException {
        String where = dummy.getState().k() + "=?";
        String whatElse = "group by " + dummy.getCertId().k() + "," + dummy.getServiceUuid().k();
        if (limit != null)
            whatElse += " limit " + limit;
        List<DbTransferDetails> result = sqlQueries.load(ISQLQueries.columns(dummy.getCertId(), dummy.getServiceUuid()), dummy, where, ISQLQueries.whereArgs(TransferState.NOT_STARTED), whatElse);
        return result;
    }

    public SQLResource<DbTransferDetails> getTransferResource() {
        return null;
    }


    public List<DbTransferDetails> getNotStartedTransfers(Long certId, String serviceUuid, int limit) throws SqlQueriesException {
        String where = dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and " + dummy.getState().k() + "=? and " + dummy.getDeleted().k() + "=? and " + dummy.getAvailable().k() + "=? ";
        String whatElse = " limit ?";
        List<DbTransferDetails> result = sqlQueries.load(dummy.getAllAttributes(), dummy, where, ISQLQueries.whereArgs(certId, serviceUuid, TransferState.NOT_STARTED, false, true, limit), whatElse);
        return result;
    }

    public void deleteByHash(String hash) throws SqlQueriesException {
        if (hash == null)
            return;
        sqlQueries.delete(dummy, dummy.getHash().k() + "=?", ISQLQueries.whereArgs(hash));
    }

    public void updateState(Long id, TransferState state) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getState().k() + "=? where " + dummy.getId().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(state, id));
    }

//    private void setStarted(Long id, boolean started) throws SqlQueriesException {
//        TransferDetails details = new TransferDetails();
//        String stmt = "update " + details.getTableName() + " set " + details.getState().k() + "=? where " + details.getId().k() + "=?";
//        sqlQueries.execute(stmt, ISQLQueries.whereArgs(true, id));
//    }

    public boolean hasNotStartedTransfers(Long certId, String serviceUuid) throws SqlQueriesException {
        String query = "select count(*)>0 from " + dummy.getTableName() + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=?";
        Integer result = sqlQueries.queryValue(query, Integer.class, ISQLQueries.whereArgs(certId, serviceUuid));
        return SqliteConverter.intToBoolean(result);
    }

    public void resetStarted() throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getState().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(TransferState.NOT_STARTED));
    }

    public Long count(Long certId, String serviceUuid) throws SqlQueriesException {
        String query = "select count (*) from " + dummy.getTableName() + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=?";
        return sqlQueries.queryValue(query, Long.class, ISQLQueries.whereArgs(certId, serviceUuid));
    }

    public Long countDone(Long certId, String serviceUuid) throws SqlQueriesException {
        String query = "select count (*) from " + dummy.getTableName()
                + " where " + dummy.getCertId().k() + "=? and "
                + dummy.getServiceUuid().k() + "=? and "
                + dummy.getState().k() + "=?";
        return sqlQueries.queryValue(query, Long.class, ISQLQueries.whereArgs(certId, serviceUuid, TransferState.DONE));
    }

    public int countStarted(Long certId, String serviceUuid) throws SqlQueriesException {
        String query = "select count (*) from " + dummy.getTableName() + " where " + dummy.getState().k() + "=? and "
                + dummy.getCertId().k() + "=? and "
                + dummy.getServiceUuid().k() + "=?";
        return sqlQueries.queryValue(query, Integer.class, ISQLQueries.whereArgs(true, certId, serviceUuid));
    }

    public void updateTransferredBytes(Long id, Long transferred) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getTransferred().k() + "=? where " + dummy.getId().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(transferred, id));
    }

    public ISQLResource<DbTransferDetails> getUnnecessaryTransfers() throws SqlQueriesException {
        FsFile f = new FsFile();
        String query = "select * from " + dummy.getTableName() + " t where not exists (select * from " + f.getTableName()
                + " f where f." + f.getContentHash().k() + "=t." + dummy.getHash().k() + " and " + f.getSynced().k() + "=?)";
        return sqlQueries.loadQueryResource(query, dummy.getAllAttributes(), DbTransferDetails.class, ISQLQueries.whereArgs(false));
    }

    public void flagDeleted(Long transferId, boolean flag) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getDeleted().k() + "=? where " + dummy.getId().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(flag, transferId));
    }

    public void updateAvailableByHashSet(Long certId, String serviceUUID, Collection<String> hashes) throws SqlQueriesException {
        String statement = "update " + dummy.getTableName() + " set " + dummy.getAvailable().k() + "=? where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and "
                + dummy.getHash().k() + " in ";
        statement += ISQLQueries.buildPartIn(hashes);
        List<Object> whereArgs = ISQLQueries.whereArgs(true, certId, serviceUUID);
        whereArgs.addAll(hashes);
        sqlQueries.execute(statement, whereArgs);
    }

    public ISQLResource<DbTransferDetails> getNotStartedNotAvailableTransfers(Long certId) throws SqlQueriesException {
        FsFile f = new FsFile();
        List<Pair<?>> attributes = new ArrayList<>();
        attributes.add(dummy.getHash());
        String query = "select t." + dummy.getHash().k() + " from " + dummy.getTableName() + " t left join " + f.getTableName() + " f on f." + f.getContentHash().k() + " = t." + dummy.getHash().k() + " " +
                "where t." + dummy.getState().k() + "=? and t." + dummy.getAvailable().k() + "=? and t." + dummy.getCertId().k() + "=? group by t." + dummy.getHash().k();
        ISQLResource<DbTransferDetails> result = sqlQueries.loadQueryResource(query, attributes, DbTransferDetails.class, ISQLQueries.whereArgs(false, false, certId));
        return result;
    }


    public synchronized ISQLResource<FsFile> getAvailableTransfersFromHashes(Iterator<AvailHashEntry> iterator) throws SqlQueriesException {
        /**
         * this shovels all hashes into a temporary table and joins it with the fs table.
         */
        MissingHash m = new MissingHash();
        FsFile f = new FsFile();
        sqlQueries.delete(m, null, null);
        N.forEach(iterator, availHashEntry -> {
            this.insertHashTmp(availHashEntry.getHash());
        });
        List<Pair<?>> attributes = new ArrayList<>();
        attributes.add(f.getContentHash());
        String query = "select f." + f.getContentHash().k() + " from " + m.getTableName() + " m left join " + f.getTableName() + " f on m." + m.getHash().k() + "=" + f.getContentHash().k()
                + " where f." + f.getSynced().k() + "=? and f." + f.getIsDirectory().k() + "=?";
        ISQLResource<FsFile> resource = sqlQueries.loadQueryResource(query, attributes, FsFile.class, ISQLQueries.whereArgs(true, false));
        return resource;
    }

    /**
     * insert hashes to temporary table
     *
     * @param hash
     */
    private void insertHashTmp(String hash) throws SqlQueriesException {
        sqlQueries.insert(new MissingHash(hash));
    }

    public void flagNotStartedHashAvailable(String hash) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getAvailable().k() + "=? where " + dummy.getState().k() + "=? and " + dummy.getHash().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(true, false, hash));
    }

    public void updateStateByHash(String hash, TransferState state) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName()
                + " set " + dummy.getState().k() + "=? where " + dummy.getHash().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(state, hash));
    }

    public void deleteDone(@NotNull Long partnerCertificateId, @NotNull String partnerServiceUuid) throws SqlQueriesException {
        String stmt = "delete from " + dummy.getTableName() + " where "
                + dummy.getCertId().k() + "=? and "
                + dummy.getServiceUuid().k() + "=? and "
                + dummy.getState().k() + "=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(partnerCertificateId, partnerServiceUuid, TransferState.DONE));
    }

    public void flagStateForRemainingTransfers(long certId, @NotNull String serviceuuid, TransferState state) throws SqlQueriesException {
        String stmt = "update " + dummy.getTableName() + " set " + dummy.getState().k() + "=?"
                + " where " + dummy.getCertId().k() + "=? and " + dummy.getServiceUuid().k() + "=? and " + dummy.getState().k() + "!=? and " + dummy.getState().k() + "!=?";
        sqlQueries.execute(stmt, ISQLQueries.whereArgs(state, certId, serviceuuid, TransferState.DONE, TransferState.RUNNING));
    }


    public boolean hasHash(String hash) throws SqlQueriesException {
        String query = "select count(1) from " + dummy.getTableName() + " where " + dummy.getHash().k() + "=?";
        Long count = sqlQueries.queryValue(query, Long.class, ISQLQueries.whereArgs(hash));
        return count > 0;
    }
}

File: filesync/src/main/java/de/mel/drive/sql/dao/TransferDao.kt
package de.mel.drive.sql.dao

import de.mel.auth.data.db.MissingHash
import de.mel.auth.tools.N
import de.mel.drive.sql.*
import de.mel.drive.tasks.AvailHashEntry
import de.mel.sql.*
import java.util.*

/**
 * Created by xor on 12/16/16.
 */
class TransferDao : Dao {

    companion object {
        /**
         * this method allows us to remove the dependency to sqlite.
         * This saves more than a megabyte on Android.
         */
        private fun violatesUnique(e: Exception): Boolean {
            if (e.javaClass.name == "org.sqlite.SQLiteException") {
                return e.message?.contains("(UNIQUE constraint") ?: false
            }
            return false
        }
    }

    private val dummy = DbTransferDetails()

    val oneTransfer: DbTransferDetails?
        @Throws(SqlQueriesException::class)
        get() {

            val res = sqlQueries.load(dummy.allAttributes, dummy, null, null, " order by " + dummy.id.k() + " limit 1")
            return if (res.size == 1) res[0] else null
        }

    val twoTransferSets: MutableList<DbTransferDetails>
        @Throws(SqlQueriesException::class)
        get() = getTransferSets(2)

    val transferResource: SQLResource<DbTransferDetails>?
        get() = null

    val unnecessaryTransfers: ISQLResource<DbTransferDetails>
        @Throws(SqlQueriesException::class)
        get() {
            val f = FsFile()
            val query = ("select * from " + dummy.tableName + " t where not exists (select * from " + f.tableName
                    + " f where f." + f.contentHash.k() + "=t." + dummy.hash.k() + " and " + f.synced.k() + "=?)")
            return sqlQueries.loadQueryResource(query, dummy.allAttributes, DbTransferDetails::class.java, ISQLQueries.args(false))
        }

    constructor(ISQLQueries: ISQLQueries) : super(ISQLQueries) {}

    constructor(ISQLQueries: ISQLQueries, lock: Boolean) : super(ISQLQueries, lock) {}

    @Throws(SqlQueriesException::class)
    fun getLeftoversByService(certId: Long?, serviceUuid: String): TransferLeftovers? {
        val lo = TransferLeftovers()
//        val query = ("select (select count(1) from " + dummy.tableName + ") as " + lo.filesTotal.k()
//                + ", (select count(1) from " + dummy.tableName + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and " + dummy.state.k() + "=? ) as " + lo.filesTransferred.k()
//                + ", select( sum(" + dummy.size.k() + ") as " + lo.bytesTotal.k()
//                + ", (select sum(" + dummy.transferred.k() + ") from " + dummy.tableName + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and " + dummy.state.k() + " != ?) as " + lo.bytesTransferred.k()
//                + " from " + dummy.tableName + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and " + dummy.state.k() + " !=?")

        val query = "select count(1) as ${lo.filesTotal.k()}" +
                ", (select count(1) from ${dummy.tableName} where ${dummy.certId.k()}=? and ${dummy.serviceUuid.k()}=? and ${dummy.state.k()}=?) as ${lo.filesTransferred.k()}" +
                ", sum(${dummy.size.k()}) as ${lo.bytesTotal.k()}" +
                ", sum(${dummy.transferred.k()}) as ${lo.bytesTransferred.k()}" +
                " from ${dummy.tableName} where ${dummy.certId.k()}=? and ${dummy.serviceUuid.k()}=?"
        val list = sqlQueries.loadString(lo.allAttributes, lo, query, ISQLQueries.args(certId, serviceUuid, TransferState.DONE, certId, serviceUuid))
        if (!list.isEmpty()) {
            // summed bytes are null if no transfer exists
            val result = list[0]
            if (result.bytesTransferred.equalsValue(null))
                result.bytesTransferred.v(0L)
            if (result.bytesTotal.equalsValue(null))
                result.bytesTotal.v(0L)
            return result
        }
        return null
    }

    @Throws(SqlQueriesException::class)
    fun insert(dbTransferDetails: DbTransferDetails) {
        try {
            val id = sqlQueries.insert(dbTransferDetails)
            dbTransferDetails.id.v(id)
        } catch (sqlQueriesException: SqlQueriesException) {
            // non unique transfers happen quite often and thats normal. don't fill the logs with that.
            if (!violatesUnique(sqlQueriesException.exception))
                throw sqlQueriesException
        }
    }

    @Throws(SqlQueriesException::class)
    fun delete(id: Long?) {

        sqlQueries.delete(dummy, dummy.id.k() + "=?", ISQLQueries.args(id))
    }

    /**
     * @param limit
     * @return all TransferSets if limit is null
     */
    @Throws(SqlQueriesException::class)
    fun getTransferSets(limit: Int?): MutableList<DbTransferDetails> {
        val where = dummy.state.k() + "=?"
        var whatElse = "group by " + dummy.certId.k() + "," + dummy.serviceUuid.k()
        if (limit != null)
            whatElse += " limit $limit"
        return sqlQueries.load(ISQLQueries.columns(dummy.certId, dummy.serviceUuid), dummy, where, ISQLQueries.args(TransferState.NOT_STARTED), whatElse)
    }


    @Throws(SqlQueriesException::class)
    fun getNotStartedTransfers(certId: Long?, serviceUuid: String, limit: Int): MutableList<DbTransferDetails> {
        val where = dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and " + dummy.state.k() + "=? and " + dummy.deleted.k() + "=? and " + dummy.available.k() + "=? "
        val whatElse = " limit ?"
        return sqlQueries.load(dummy.allAttributes, dummy, where, ISQLQueries.args(certId, serviceUuid, TransferState.NOT_STARTED, false, true, limit), whatElse)
    }

    @Throws(SqlQueriesException::class)
    fun deleteByHash(hash: String?) {
        if (hash == null)
            return
        sqlQueries.delete(dummy, dummy.hash.k() + "=?", ISQLQueries.args(hash))
    }

    @Throws(SqlQueriesException::class)
    fun updateState(id: Long?, state: TransferState) {
        val stmt = "update " + dummy.tableName + " set " + dummy.state.k() + "=? where " + dummy.id.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(state, id))
    }

    //    private void setStarted(Long id, boolean started) throws SqlQueriesException {
    //        TransferDetails details = new TransferDetails();
    //        String stmt = "update " + details.getTableName() + " set " + details.getState().k() + "=? where " + details.getId().k() + "=?";
    //        sqlQueries.execute(stmt, ISQLQueries.whereArgs(true, id));
    //    }

    @Throws(SqlQueriesException::class)
    fun hasNotStartedTransfers(certId: Long?, serviceUuid: String): Boolean {
        val query = "select count(*)>0 from " + dummy.tableName + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=?"
        val result = sqlQueries.queryValue(query, Int::class.java, ISQLQueries.args(certId, serviceUuid))
        return SqliteConverter.intToBoolean(result)!!
    }

    @Throws(SqlQueriesException::class)
    fun resetStarted() {
        val stmt = "update " + dummy.tableName + " set " + dummy.state.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(TransferState.NOT_STARTED))
    }

    @Throws(SqlQueriesException::class)
    fun count(certId: Long?, serviceUuid: String): Long {
        val query = "select count (*) from " + dummy.tableName + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=?"
        return sqlQueries.queryValue(query, Long::class.java, ISQLQueries.args(certId, serviceUuid))
    }

    @Throws(SqlQueriesException::class)
    fun countRemaining(certId: Long?, serviceUuid: String): Long {
        val query = "select count (*) from ${dummy.tableName} where ${dummy.certId.k()}=? and ${dummy.serviceUuid.k()}=? " +
                "and ${dummy.state.k()}=?"
        return sqlQueries.queryValue(query, Long::class.java, ISQLQueries.args(certId, serviceUuid, TransferState.NOT_STARTED))
    }

    @Throws(SqlQueriesException::class)
    fun countStarted(certId: Long?, serviceUuid: String): Int {
        val query = ("select count (*) from " + dummy.tableName + " where " + dummy.state.k() + "=? and "
                + dummy.certId.k() + "=? and "
                + dummy.serviceUuid.k() + "=?")
        return sqlQueries.queryValue(query, Int::class.java, ISQLQueries.args(true, certId, serviceUuid))
    }

    @Throws(SqlQueriesException::class)
    fun updateTransferredBytes(id: Long?, transferred: Long?) {
        val stmt = "update " + dummy.tableName + " set " + dummy.transferred.k() + "=? where " + dummy.id.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(transferred, id))
    }

    @Throws(SqlQueriesException::class)
    fun flagForDeletion(transferId: Long?, flag: Boolean) {
        val stmt = "update " + dummy.tableName + " set " + dummy.deleted.k() + "=? where " + dummy.id.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(flag, transferId))
    }

    @Throws(SqlQueriesException::class)
    fun updateAvailableByHashSet(certId: Long?, serviceUUID: String, hashes: Collection<String>) {
        var statement = ("update " + dummy.tableName + " set " + dummy.available.k() + "=? where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and "
                + dummy.hash.k() + " in ")
        statement += ISQLQueries.buildPartIn(hashes)
        val whereArgs = ISQLQueries.args(true, certId, serviceUUID)
        whereArgs.addAll(hashes)
        sqlQueries.execute(statement, whereArgs)
    }

    @Throws(SqlQueriesException::class)
    fun getNotStartedNotAvailableTransfers(certId: Long?): ISQLResource<DbTransferDetails> {
        val f = FsFile()
        val attributes = ArrayList<Pair<*>>()
        attributes.add(dummy.hash)
        val query = "select t." + dummy.hash.k() + " from " + dummy.tableName + " t left join " + f.tableName + " f on f." + f.contentHash.k() + " = t." + dummy.hash.k() + " " +
                "where t." + dummy.state.k() + "=? and t." + dummy.available.k() + "=? and t." + dummy.certId.k() + "=? group by t." + dummy.hash.k()
        return sqlQueries.loadQueryResource(query, attributes, DbTransferDetails::class.java, ISQLQueries.args(false, false, certId))
    }


    @Synchronized
    @Throws(SqlQueriesException::class)
    fun getAvailableTransfersFromHashes(iterator: Iterator<AvailHashEntry>): ISQLResource<FsFile> {
        /**
         * this shovels all hashes into a temporary table and joins it with the fs table.
         */
        val m = MissingHash()
        val f = FsFile()
        sqlQueries.delete(m, null, null)
        N.forEach(iterator) { availHashEntry -> this.insertHashTmp(availHashEntry.hash) }
        val attributes = ArrayList<Pair<*>>()
        attributes.add(f.contentHash)
        val query = ("select f." + f.contentHash.k() + " from " + m.tableName + " m left join " + f.tableName + " f on m." + m.hash.k() + "=" + f.contentHash.k()
                + " where f." + f.synced.k() + "=? and f." + f.isDirectory.k() + "=?")
        return sqlQueries.loadQueryResource(query, attributes, FsFile::class.java, ISQLQueries.args(true, false))
    }

    /**
     * insert hashes to temporary table
     *
     * @param hash
     */
    @Throws(SqlQueriesException::class)
    private fun insertHashTmp(hash: String) {
        sqlQueries.insert(MissingHash(hash))
    }

    @Throws(SqlQueriesException::class)
    fun flagNotStartedHashAvailable(hash: String) {
        val stmt = "update " + dummy.tableName + " set " + dummy.available.k() + "=? where " + dummy.state.k() + "=? and " + dummy.hash.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(true, false, hash))
    }

    @Throws(SqlQueriesException::class)
    fun updateStateByHash(hash: String, state: TransferState) {
        val stmt = ("update " + dummy.tableName
                + " set " + dummy.state.k() + "=? where " + dummy.hash.k() + "=?")
        sqlQueries.execute(stmt, ISQLQueries.args(state, hash))
    }

    //    public void deleteDone(@NotNull Long partnerCertificateId, @NotNull String partnerServiceUuid) throws SqlQueriesException {
    //        String stmt = "delete from " + dummy.getTableName() + " where "
    //                + dummy.getCertId().k() + "=? and "
    //                + dummy.getServiceUuid().k() + "=? and "
    //                + dummy.getState().k() + "=?";
    //        sqlQueries.execute(stmt, ISQLQueries.whereArgs(partnerCertificateId, partnerServiceUuid, TransferState.DONE));
    //    }

    @Throws(SqlQueriesException::class)
    fun flagStateForRemainingTransfers(certId: Long, serviceuuid: String, state: TransferState) {
        val stmt = ("update " + dummy.tableName + " set " + dummy.state.k() + "=?"
                + " where " + dummy.certId.k() + "=? and " + dummy.serviceUuid.k() + "=? and " + dummy.state.k() + "!=? and " + dummy.state.k() + "!=?")
        sqlQueries.execute(stmt, ISQLQueries.args(state, certId, serviceuuid, TransferState.DONE, TransferState.RUNNING))
    }


    @Throws(SqlQueriesException::class)
    fun hasHash(hash: String): Boolean {
        val query = "select count(1) from " + dummy.tableName + " where " + dummy.hash.k() + "=?"
        val count = sqlQueries.queryValue(query, Long::class.java, ISQLQueries.args(hash))
        return count > 0L
    }

    @Throws(SqlQueriesException::class)
    fun deleteFlaggedForDeletion() {
        val stmt = "delete from " + dummy.tableName + " where " + dummy.deleted.k() + "=?"
        sqlQueries.execute(stmt, ISQLQueries.args(true))
    }

    fun flagForDeletionByHash(hash: String) {
        val stmt = "update ${dummy.tableName} set ${dummy.deleted.k()}=?, ${dummy.state.k()}=? where ${dummy.hash.k()}=?"
        sqlQueries.execute(stmt, ISQLQueries.args(true, TransferState.DONE, hash))
    }
}
